### PR TITLE
Support creation of Node subclasses

### PR DESCRIPTION
**Pull request by wayneparrott**
Added ability to create Node subclasses. 
This PR maintains the public 0.17 api.

#753

### CODE REVIEW

File: lib/context.js [hunk 1]
```diff
@@ -19,64 +19,201 @@ const rclnodejs = require('bindings')('rclnodejs');
...
+ *                                     |
+ *         ----------------------------
+ *        |
+ *        v
+ * rcl.shutdown(context)
+ *        or
+ * context.shutdown() ---> shutdown
+ * ```
+ * Must call rclnodejs.init(context) to initialize the context
+ * to the usable 'initialized' (valid) state be using.
```
Comment (minggangw) [inline lib/context.js:44]: Thanks for adding this state transition diagram.

File: lib/context.js [hunk 2]
```diff
@@ -19,64 +19,201 @@ const rclnodejs = require('bindings')('rclnodejs');
...
+   * @returns {undefined}
+   * @throws {Error} If there is a problem shutting down the context.
+   */
+  tryShutdown() {
+    if (this.isValid()) {
+      this.shutdown();
+    }
+  }
+
+  onNodeCreate(node) {
```
Comment (minggangw) [inline lib/context.js]: nit: how about `onNodeCreated`?

File: lib/context.js [hunk 3]
```diff
@@ -19,64 +19,201 @@ const rclnodejs = require('bindings')('rclnodejs');
...
+
+    if (this._nodes.includes(node)) {
+      // do nothing
+      return;
+    }
+
+    this._nodes.push(node);
+  }
+
+  onNodeDestroy(node) {
```
Comment (minggangw) [inline lib/context.js]: nit: how about `onNodeDestroyed`?

**Review by minggangw — APPROVED** [review]
Thanks for your PR, good practice of OOP and we should have had it earlier. LGTM!

**Comment by wayneparrott** [comment]
@minggangw Thx for the feedback. 
I wanted to get early positive feedback on the general idea of this PR before doing the extra work of expanding our test suite to exercise the full api changes. A big TODO is for the PR to include updates the test suite to cover the Node and Context api changes before we consider the PR ready. (Perhaps I should have created this PR in draft status.)

Also I want to focus some attention and get feedback on the tagging of some rcl functions @deprecated such as createNode(), createLifecycleNode(), spin(), that are now public methods on Node. I'm not opposed to removing the @deprecated tag if we feel OK with supporting the current functional api approach and this newer object-oriented api approach going forward. Thoughts?

**Comment by minggangw** [comment]
It seems that the [Last successful build (#2125)](https://ci.ros2.org/view/packaging/job/packaging_linux/) has some problem and the ci fails constantly, I am going to re-trigger it once a new packaging is generated.

**Comment by minggangw** [comment]
Would you please take a look at the error reported by [lint](https://travis-ci.org/github/RobotWebTools/rclnodejs/builds/750445596#L25279)

```bash
/root/rclnodejs/types/node.d.ts
  121:1  error  Replace `⏎` with `··`  prettier/prettier
  123:1  error  Insert `··`            prettier/prettier
  124:4  error  Insert `··`            prettier/prettier
  125:1  error  Insert `··`            prettier/prettier
  126:4  error  Insert `··`            prettier/prettier
  127:1  error  Insert `··`            prettier/prettier
  128:4  error  Insert `··`            prettier/prettier
  129:1  error  Insert `··`            prettier/prettier
```

**Comment by minggangw** [comment]
Merged, thanks for your effort!

**Pull request merged**